executor/builder.go

```diff
@@ -2992,6 +2992,22 @@ func newRowDecoder(ctx sessionctx.Context, schema *expression.Schema, tbl *model
 			Ft:            col.RetType,
 		}
 	}
+	if len(pkCols) == 0 {
+		if tbl.IsCommonHandle {
+			var pkIdx *model.IndexInfo
+			for _, idx := range tbl.Indices {
+				if idx.Primary {
+					pkIdx = idx
+					break
+				}
+			}
```

Comment:
Could we use a function to do this, so that avoid coping the code 3 times in different place.